### PR TITLE
From pretrained library

### PR DESCRIPTION
**Pull request by sordonia**
* Adds `MultiExpertModel.from_pretrained_library("....", selector_config=ArrowSelectorConfig())`

* Every selector (PG, Hidden, Arrow) can now load their own prototypes from the library automatically

### CODE REVIEW

File: mttl/models/containers/selectors/base_selectors.py [hunk 1]
```diff
@@ -1,16 +1,15 @@
...
-import wandb
 import numpy as np
 from torch import nn
 import torch.nn.functional as F
 from mttl.models.library.expert import ExpertInfo
-from mttl.models.modifiers.routing import RoutingInfo
-from torch.distributions import Bernoulli, Categorical
-from mttl.models.utils import MetricLogger
+from torch.distributions import Categorical
+from mttl.models.utils import MetricLogger, Singleton
```
Comment (pclucas14) [inline mttl/models/containers/selectors/base_selectors.py:12]: where is Singleton used?

File: mttl/models/containers/selectors/arrow_selector.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+from mttl.models.containers.selectors import (
+    PerTokenSelector,
+    PerTokenSelectorConfig,
+    register_multi_expert_selector,
+)
+from mttl.models.library.expert_library import ExpertLibrary
+from mttl.models.library.library_transforms import ArrowConfig, ArrowTransform
+
+
+def get_arrow_embeddings(
```
Comment (pclucas14) [inline mttl/models/containers/selectors/arrow_selector.py:15]: I would name `compute_arrow_embeddings` as we are not actually getting the embeddings (only their save name)

Comment (sordonia) [reply mttl/models/containers/selectors/arrow_selector.py]: fixed

File: mttl/models/containers/selectors/base_selectors.py [hunk 3]
```diff
@@ -813,219 +734,21 @@ def forward(self, input, **kwargs) -> BatchSequenceExpertsAndWeightsSelectorOutp
...
-
-        if "layer" not in kwargs:
-            raise ValueError(
-                "PhatgooseSelector requires a layer to be passed in kwargs to infer the input dimension."
+        if self.library_artifacts is not None:
+            proto = get_expert_prototype_from_library_artifacts(
+                expert_name, self.layer_name, self.library_artifacts
+            ).unsqueeze(0)
+        else:
+            proto = torch.zeros(
```
[the remark this answers is not in the document]
Comment (sordonia) [reply mttl/models/containers/selectors/base_selectors.py:747]: ok good point

File: mttl/models/containers/selectors/arrow_selector.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
+from dataclasses import dataclass
+
+from mttl.models.containers.selectors.base_selectors import (
+    PerTokenSelector,
+    PerTokenSelectorConfig,
+    register_multi_expert_selector,
+)
+
+
+def comput_arrow_embeddings(
```
Comment (pclucas14) [inline mttl/models/containers/selectors/arrow_selector.py:10]: typo

File: mttl/models/containers/selectors/arrow_selector.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+) -> str:
+    from mttl.models.library.library_transforms import ArrowConfig, ArrowTransform
+
+    cfg = ArrowConfig(
+        name=selector_data_id,
+        ab_only=ab_only,
+        tie_params=tie_params or "default",
+        tie_op=tie_op,
+    )
+    ArrowTransform(cfg).transform(
```
Comment (pclucas14) [inline mttl/models/containers/selectors/arrow_selector.py:27]: we should probably hardcode `persist=True`

File: mttl/models/containers/selectors/average_activation_selector.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        HiddenStateComputer,
+        HiddenStateComputerConfig,
+    )
+
+    cfg = HiddenStateComputerConfig(
+        name=selector_data_id,
+        use_base_model_only=use_base_model_only,
+        max_samples_per_task=max_samples_per_task,
+        track=track,
+        pool=pool,
```
Comment (pclucas14) [inline mttl/models/containers/selectors/average_activation_selector.py:30]: `persist=True`

File: mttl/models/containers/selectors/phatgoose_selector.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+    cfg = PhatgooseConfig(
+        n_steps=n_steps_pg,
+        learning_rate=learning_rate_pg,
+        name=selector_data_id,
+    )
+
+    phatgoose_transform = PhatgooseTransform(cfg)
+    phatgoose_transform.transform(
+        library, default_args=default_args, recompute=recompute_prototypes
```
Comment (pclucas14) [inline mttl/models/containers/selectors/phatgoose_selector.py:41]: `persist=True`

File: projects/modular_llm/eval_library.py [hunk 8]
```diff
@@ -242,13 +214,15 @@ def run_eval(args: ExpertConfig):
...
+        if not args.selector_data_id:
+            args.selector_data_id = fetch_prototypes(args, library)
 
         module = MultiExpertModel(
             **vars(train_cfg), selector_config=selector_config
         ).to("cuda")
 
         module.add_experts_from_library(library)
-        patch_prototypes(module, library, args)
+        # patch_prototypes(module, library, args)
```
Comment (matheper) [inline projects/modular_llm/eval_library.py:225]: Does `patch_prototypes` still exists?
`mttl/models/expert_model.py` is also referring to it:
```
from projects.modular_llm.eval_library import patch_prototypes
patch_prototypes(self, expert_library, self.selector_config)
```

File: projects/modular_llm/eval_library.py [hunk 9]
```diff
@@ -201,7 +173,7 @@ def run_eval(args: ExpertConfig):
     )
     an_expert = library[next(iter(library.keys()))]
     train_cfg = deepcopy(an_expert.training_config)
-    train_cfg.device_map = "cpu"
+    # train_cfg.device_map = "cpu"
```
Comment (matheper) [inline projects/modular_llm/eval_library.py:176]: what about removing this line?

File: mttl/models/containers/selectors/arrow_selector.py (hunk 4 above)
Comment (sordonia) [reply mttl/models/containers/selectors/arrow_selector.py]: solved

File: mttl/models/containers/selectors/arrow_selector.py (hunk 5 above)
Comment (sordonia) [reply mttl/models/containers/selectors/arrow_selector.py:27]: ok good idea

File: projects/modular_llm/eval_library.py (hunk 8 above)
Comment (sordonia) [reply projects/modular_llm/eval_library.py:225]: good call

File: projects/modular_llm/eval_library.py (hunk 9 above)
Comment (sordonia) [reply projects/modular_llm/eval_library.py:176]: ok

File: mttl/models/utils.py [hunk 10]
```diff
@@ -20,6 +20,13 @@
 CHECKPOINT_PATH_IN_HUB = "checkpoint.ckpt"
 
 
+class Singleton(object):
```
Comment (pclucas14) [inline mttl/models/utils.py:23]: where is this used ?

**Review by pclucas14 — COMMENTED** [review]
🚀

File: mttl/models/utils.py [hunk 11]
```diff
@@ -447,11 +454,12 @@ def model_loader_helper(
         )
     elif "phi-2" == model_name:
         # local phi-2 version. use `microsoft/phi-2 for the official hf version`
-        logger.info(
-            f"Loading phi-2 model from {os.getenv('PHI_PATH', 'microsoft/phi-2')}"
-        )
+        if "PHI_PATH" not in os.environ:
```
Comment (matheper) [inline mttl/models/utils.py:457]: If I remember correctly, the logic here was to not break configs using `phi-2`. If there is no `PHI_PATH` we default to huggingface.

File: tests/test_expert_model.py [hunk 12]
```diff
@@ -45,7 +46,71 @@ def test_expert_model():
         model.get_merged_expert()
 
     assert len(model.selectors["lora"]) == 12
-    assert isinstance(next(iter(model.selectors["lora"].values())), TaskNameSelector)
+    assert isinstance(model.selectors["lora"][0], TaskNameSelector)
+
+
+def test_from_pretrained():
+    import tempfile
```
Comment (matheper) [inline tests/test_expert_model.py:53]: why not `tmp_path`?

`def test_from_pretrained(tmp_path):`

File: mttl/models/utils.py (hunk 11 above)
Comment (sordonia) [reply mttl/models/utils.py:457]: ok

Comment (sordonia) [reply mttl/models/utils.py:457]: but actually if we default to microsoft/phi-2 and use our phi private library the code will error out with a weird error

**Review by matheper — APPROVED** [review]

**Pull request merged**